infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -186,6 +209,66 @@ private void initializeLogMetadata() {
         log.info("initializeStreamTails: took {} ms to load {}, log start {}", end - start, logMetadata, getTrimMark());
     }
 
+    /**
+     * Load log metadata (stream address space map and stream tail map) from
+     * persisted log metadata file for fast startup. Only addresses in range
+     * [startingAddress, maxAddress] are considered valid and are loaded.
+     * @param maxAddress load addresses up to this address (inclusive)
+     * @return the highest stream tail loaded
+     */
+    private Long loadPersistedLogMetadata(long maxAddress) {
+        long start = System.currentTimeMillis();
+        Map<UUID, List<Long>> streamAddressSpaceMap = sc.getLogUnitMetadata();
+        if (streamAddressSpaceMap == null) {
+            log.info("Unable to load persisted log metadata due to missing file.");
+            return Address.NON_ADDRESS;
+        }
+
+        AtomicLong numOfAddrLoaded = new AtomicLong(0L);
+        streamAddressSpaceMap.forEach((uuid, addresses) -> {
+            Set<Long> validAddresses = addresses.stream()
+                    .filter(address -> address >= dataStore.getStartingAddress() &&
+                            address <= maxAddress)
+                    .collect(Collectors.toSet());
+            logMetadata.getStreamsAddressSpaceMap().put(uuid, new StreamAddressSpace(validAddresses));
+
+            long streamTail = validAddresses.stream()
+                    .mapToLong(Long::longValue).max().orElse(Address.NON_ADDRESS);
+            logMetadata.getStreamTails().put(uuid, streamTail);
+            numOfAddrLoaded.addAndGet(validAddresses.size());
+        });
+
+        long highestTail = Collections.max(logMetadata.getStreamTails().values());
+        long end = System.currentTimeMillis();
+        log.info("Loaded {} valid addresses from persisted log metadata in {} ms. Highest stream tail loaded is {}",
+                numOfAddrLoaded, end-start, highestTail);
+
+        return highestTail;
+    }
+
+    /**
+     * Persist the current stream address space map on disk.
+     */
+    @Override
+    public void persistLogMetadata() {
+        log.info("Start persisting log metadata.");
+        long start = System.currentTimeMillis();
+
```

Comment:
This also needs to persist the `StreamAddressSpace::trimMark`

Reply:
Done